### PR TITLE
Update `azurerm_linux|windows_virtual_machine_scale_set` - Fix in some circumstances updating VMSS error complaining imageReference is null

### LINKED ISSUE #7957: `azurerm_linux_virtual_machine_scale_set` - update error
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

```
Terraform v0.12.29
```

### Affected Resource(s)

* `azurerm_linux_virtual_machine_scale_set`
* probably `azurerm_windows_virtual_machine_scale_set`

### Terraform Configuration Files

```hcl
resource "azurerm_linux_virtual_machine_scale_set" "main" {
  name                            = "${var.prefix}-vmss"
  resource_group_name             = data.azurerm_resource_group.test.name
  location                        = data.azurerm_resource_group.test.location
  sku                             = "Standard_F2"
  instances                       = 3
  admin_username                  = "adminuser"
  admin_password                  = "P@ssw0rd1234!"
  disable_password_authentication = false

  source_image_id = data.azurerm_shared_image_version.test.id

  network_interface {
    name    = "example"
    primary = true

    ip_configuration {
      name      = "internal"
      primary   = true
      subnet_id = azurerm_subnet.internal.id
    }
  }

  os_disk {
    storage_account_type = "Standard_LRS"
    caching              = "ReadWrite"
  }
  
  data_disk {
    storage_account_type = "Standard_LRS"
    caching              = "ReadOnly"
    lun                  = 1
    disk_size_gb         = 20
    create_option        = "Empty"
  }

  tags = {
    "Test" = "Foo"
  }
}
```

### Panic Output

```
Error updating LLinuxinux Virtual Machine Scale Set "test-vmss" (Resource Group "test-vmss-update"): compute.VirtualMachineScaleSetsClient#Update: Failure sending request: StatusCode=400 -- Original Error: Code="InvalidParameter" Message="Required parameter 'imageReference' is missing (null)." Target="imageReference"
```

### Expected Behavior

The VMSS successfully updated

### Actual Behavior

Got an error complaining `imageReference` is `null`
```
Error updating LLinuxinux Virtual Machine Scale Set "test-vmss" (Resource Group "test-vmss-update"): compute.VirtualMachineScaleSetsClient#Update: Failure sending request: StatusCode=400 -- Original Error: Code="InvalidParameter" Message="Required parameter 'imageReference' is missing (null)." Target="imageReference"
```

### Steps to Reproduce

1. `terraform apply` with the config above but the `data_disk` block removed - the VMSS should be successfully provisioned
1. `terraform apply` again with the config above with `data_disk` block restored - service throws an error at you

### Important Factoids

I believe this should be an Azure API bug - in some circumstances, the API is expected in a PATCH request, even though you are not willing to change the image reference, you will have to specify an image reference.

But a workaround on the provider side is also valid - just always fill in the image reference in the update callback

### References

* #7253

### PR DESCRIPTION
**Pull request by ArcturusZhang**
Fixes #7957

### CODE REVIEW

File: azurerm/internal/services/compute/linux_virtual_machine_scale_set_resource.go [hunk 1]
```diff
@@ -527,8 +527,22 @@ func resourceArmLinuxVirtualMachineScaleSetUpdate(d *schema.ResourceData, meta i
 		return fmt.Errorf("Error retrieving Linux Virtual Machine Scale Set %q (Resource Group %q): `properties` was nil", id.Name, id.ResourceGroup)
 	}
 
+	sourceImageReferenceRaw := d.Get("source_image_reference").([]interface{})
+	sourceImageId := d.Get("source_image_id").(string)
+	sourceImageReference, err := expandSourceImageReference(sourceImageReferenceRaw, sourceImageId)
+	if err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (ArcturusZhang) [reply azurerm/internal/services/compute/linux_virtual_machine_scale_set_resource.go:532]: Got it

File: azurerm/internal/services/compute/linux_virtual_machine_scale_set_resource.go [hunk 2]
```diff
@@ -626,30 +640,17 @@ func resourceArmLinuxVirtualMachineScaleSetUpdate(d *schema.ResourceData, meta i
 	if d.HasChange("data_disk") || d.HasChange("os_disk") || d.HasChange("source_image_id") || d.HasChange("source_image_reference") {
 		updateInstances = true
 
-		storageProfile := &compute.VirtualMachineScaleSetUpdateStorageProfile{}
+		//storageProfile := &compute.VirtualMachineScaleSetUpdateStorageProfile{}
```
[the remark this answers is not in the document]
Comment (ArcturusZhang) [reply azurerm/internal/services/compute/linux_virtual_machine_scale_set_resource.go]: Done

**Comment by ArcturusZhang** [comment]
Hi @tombuildsstuff I have updated my code by first assign the imageReference using the result from GET response, and then assign new imageReference if `HasChange` returns true. Please have a look, thanks!

**Review by tombuildsstuff — APPROVED** [review]
LGTM 👍

**Comment by tombuildsstuff** [comment]
Tests look good 👍

**Pull request merged**